### ISSUE TITLE
off-heap state access and boost fusion

### ISSUE THREAD

**Issue by jnorthrup** [opennars/opennars#37]
I wanted to see if i could make the java-side build as simple as humanly possible for at least a benchmark cut of opennars to test a notion i have to reify beans into BytebUffer expanses projected onto slabs.

A follow on would be to utilize the slabs as a proto dialect; and build a reference u-tree spirit parser for the narsese grammars for c++ side event interrop with proto to-from java.  

https://github.com/jnorthrup/prautobeans is a maven-plugin based proto IDL parserto generate java interface proxies ( GWT Autobeans/Splittable) with bytebuffer backing stores,   a complimentary library https://github.com/jnorthrup/drunken-wallhack is able to map GWT Autobeans to zero-copy json input parsed on the fly -- with two flavors of stateful and attribute-seek-only implementations, both significantly faster than gson at large input.

the first question is: 
does mavenization fit the long-term of opennars?  

i am for maven because there's potential to bring along some portable and robust c++ artifacts using boost headers-as-libraries in maven compilation, and if any part of that needs help being cleaner i might hack it the way it needs to be upstream in maven-native-plugin if it made sense and enabled other projects to work with this one.  the motorola bluez driver project has immaculate c++ build execution under maven, it's very clean to bring over.

**Comment by automenta** [comment]
combine with utf8 implemented by int[] buffers , started here 
https://github.com/opennars/opennars/blob/volatile5/nars_util/src/main/java/nars/util/utf8/IntBuf.java
https://github.com/opennars/opennars/blob/volatile5/nars_util/src/test/java/objenome/util/IntBufTest.java

utf8 byte[] keys for terms seems to improve both memory usage and also cpu because it reduces comparison and hashing cost. an int[] implementation would add additional cpu improvements, i could not find anything like this in any java utf8 libraries though

**Comment by automenta** [comment]
public interface Term extends Cloneable, Comparable<Term>, Named<byte[]>, Termed, Serializable {

notice the byte[] in Named, this identifies the key type when a Term is used in a Bag

**Comment by jnorthrup** [comment]
https://github.com/jnorthrup/drunken-wallhack/blob/master/src/main/java/com/colinalworth/gwt/beans/shared/DummySplittable.java#L12

**Comment by automenta** [comment]
i got most of the existing Utf8 / ByteBuf code from Boon JSON parser

**Comment by jnorthrup** [comment]
i dont see anything that looks like it would move the needle on performance
and apparently backed it out as easily as you brought it in.

bbcursive is in as a placeholder. bbcursive is starting to have analogs of
the bytebuffer operations for BytebufferReader operations, which is a fancy
way of making a class to grab the ByteBuffer pointers and peek/poke
directly on barrier-free RAM.  the jury is out whether you do bypass any
barriers and syncronizers but it cant hurt if you decide to mmap a few gigs
of json assertions as atoms with pointers

**Comment by jnorthrup** [comment]
i doubt anyone is writing java code that outperforms the jdk's native
string handling for utf8.  i could be wrong, but even if they did beat the
jvm for strings, at what cost for what gains?  i beleive that picutre does
change when c++ is part of the runtime, plus or minus actual zero copy
object mapping

**Comment by automenta** [comment]
i use NALPerfLoop to test optimization as a whole. on single thread mode (multithreaded modes, if they still work, arent guaranteed results yet) i generally find < 10uSec / cycle readings on my PC.  this means ~100khz can be expected for the 2k concepts the defaults there are set. i take this to mean APPROX (for various reasons) 50 "narwaves per second"

constructing the strings is part of it. getting the hash is another although standard string handling would include this.  but the other part is re-using the buffer (arraycopy) when constructing a superterm. this seems to add up.  then comparing buffers for equivalence and compareTo, this is where i think the int[] can help. but its not going to provide as much performance as somethign more drastic that will seems like it will require being more clever than just interpreting the profiler output as only i've been doing all along

**Comment by automenta** [comment]
some easy optimizations: encode the NAL operators as special characters, they are prsently naively stored as their string reprsentation, WITH spaces: " --> " (taking 5 bytes when it could take 1)

SELF is sort of a reserved term, but it could easily be replaced with "I" saving 3 bytes on every occurence

as terms become longer we could use RLE to compress them but im not certain at what threshold this would matter if any

**Comment by automenta** [comment]
the "Stamp" is a data structure assigned for each Task to prevent cyclic reasoning. currently this uses an array of unique serial #'s, limited by a maximum size, that are compared for overlap. i think another implementation of this that can work would be a Bloom filter (ex: guava's).  It is approximate but so is the finite-length array stamp.  Another advantage of the bloom filter is bitwise operations something to consider.  Anyway I do not think the timing information should be part of the stamp, i will probably move this to Task or its own class, and also occurenceTime to its own subclass to keep it distinct (occurenceTime is not used in 'eternal' pre-NAL7 reasoning).

**Comment by jnorthrup** [comment]
i acheived 90% of netty performance duplicating the RelaxFactory
single-threaded io loop and feeding new sockets round-robin to the 8
threads for 8 cores.

netty goes to great lengths to rewrite bytebuffer and nio with heap objects
and misc.unsafe;  i think they went to a ton of effort with a  lot of
developers over a long time to get a 10% edge over known latency-inducing
DirtectByteBuffer read access and java UTF_8 for marshalling strings.

for 1xio i pulled 150% performance out of my ass going from 1 thread to 8
and simply reducing, not fully pooling bytebuffer allocations -- slabs are
still gc-tracked heap objects with no finegrained purge options once
threads start divying them up.  allocating them 64 megs + at a time reduces
pressure on the gc.  that aside, 1xio uses a few % less cpu while making
comparable TPS. better buffer and block pooling strategies could bring up a
few more %.

i beleive there is no option for NIO based SSL, only stream based SSL.
 that's why 1xio seems to be the only pure select loop driven engine of the
ones i have seen.

the rest try to accomodate adapters and abstractions to all network
scenarios, again, not just a little bit due to SSL Engine being
undocumented garbage with holes and bugs and ancient ciphers.

**Comment by jnorthrup** [comment]
oh, so, those slabs are in the bbcursive addition to nars-logic in my fork
now
ill push the branch up

**Comment by automenta** [comment]
ive hesitated to go further with multithreading an individual NAR reasoner. it can be done but if the "Default" control loop and bags are directly converted to multithreading the performance will ultimatley decrease due to synchronization, locking etc. the alternate control loops are not tested but eliminate some or all of the contention.  however multithreading at the inter-NAR level seems to be the most flexible option once we learn how to appropriately divide the work. for example, a combination of high frequency and low frequency NARs operating on different scales, sharing data in some way (I/O messages, tasks, or entire concepts). all of this is unexplored as far as i know

**Comment by automenta** [comment]
the most unique and untested of the control strategies is the 'Ant' model where each thread directly processes a concept and moving to an adjacent one (decided by some local heuristic which can approximate the nature of the global Bag). this elimnates a central bag as long as two threads don't work on the same concept at once. if activation is accumulated between cycles as it should be doing now then this should also work fine for DirectProcess (whcih should probably be renamed TaskProcess since it is like the Task-level analog of a ConceptProcess). anyway the ant model might allow large large memories with many threads https://github.com/opennars/opennars/blob/volatile5/nars_logic/src/main/java/nars/model/cycle/experimental/AntCore.java

**Comment by automenta** [comment]
another model is Solid https://github.com/opennars/opennars/blob/volatile5/nars_logic/src/main/java/nars/model/impl/Solid.java which is the more predictable of them all because it applies a uniform policy to all concepts each cycle. this could be generalized to apply to a fraction P of all concepts per cycle and then for smaller values of P it would behave like Default control cycle with a Bag

**Comment by jnorthrup** [comment]
Solid sounds like the gpu winner

**Comment by automenta** [comment]
yah but all things considered i havent found much that is vector parallelizable which could be separated from cpu logic interacting with memory.  probably the best mix would be the hybrid CPU/GPU chips which hopefully java 9 is supposed to vectorize for. until then maybe https://github.com/pcpratts/rootbeer-ex can provide something. 

we have been wondering if it would be worth it to use FPGAs to run NARS-customized JVM cores in parallel.  several JVM-sized cores should fit on a commodity FPGA already.  google "jvm fpga" for some papers and http://www.jopdesign.com/ . but i dont think we've hit the limit yet for CPU on a single reasoner, combined with the potential for inter-NAR architectures for distributed processing (which i would like us to solve for p2p mesh first to guarantee mutually assured dissolution of authority, and not the easier cluster scenarios that a consolidated power would likely amass)

iirc, dr. wang has briefly described hardware implementations of NARS in some papers.  in some ways, each "Concept" can be considered a mini or virtual processor.

**Comment by automenta** [comment]
but in practice we'll likely want to save GPU and some cpu threads for external processes which NARS can invoke externally without its own micro-managing. eventually to include compiled procedures generated as a product of reasoning

https://github.com/opennars/opennars/blob/volatile5/nars_lab/src/main/java/nars/rl/lstm/SimpleLSTM.java
https://github.com/opennars/opennars/blob/volatile5/nars_util/src/main/java/nars/util/signal/Autoencoder.java
https://github.com/opennars/opennars/blob/volatile5/nars_lab/src/main/java/nars/rl/AbstractHaiQBrain.java

**Comment by PtrMan** [comment]
Some GPU assistance for nars would be a great investment because not everybody has idling FPGA's laying around :D
But yes, the GPU/CPU/whatever utilisation is a tradeoff.
I think that the usual NAL processing core logic don't map that well to an GPU, mainly because of either memory allocation or latency, we can still experiment around.
We just can experiment with GPU's for now...

**Comment by jnorthrup** [comment]
aparapi is passable for spot optimizations.

what interests me is an excuse to play with Haswell Transactional locking
instructions though i bought one that might be microcode-disabled now.
 this would favor Ant model i think to keep cores in-context processing
streams

what also seems like an outcome over some pretty far mountains is to port
nars to STL, for which i think alignment is pretty high.  having c++
mementos opens up the door for amp c++ and openal compilation models for
transactional memory and gpu code emitters.

**Comment by jnorthrup** [comment]
On Fri, May 29, 2015 at 11:18 AM, james northrup <northrup.james@gmail.com>
wrote:

> openal


s/openal/openacc/

Compiler support[edit
<http://en.wikipedia.org/w/index.php?title=OpenACC&action=edit&section=1>]

Support of OpenACC is available in commercial compilers from PGI (from
version 12.6), Cray, and CAPS.[7]
<http://en.wikipedia.org/wiki/OpenACC#cite_note-isc12-7>[9]
<http://en.wikipedia.org/wiki/OpenACC#cite_note-9>[10]
<http://en.wikipedia.org/wiki/OpenACC#cite_note-10>

OpenUH[11] <http://en.wikipedia.org/wiki/OpenACC#cite_note-11> is an Open64
<http://en.wikipedia.org/wiki/Open64> based open source OpenACC compiler
supporting C and FORTRAN, developed by HPCTools group from University of
Houston <http://en.wikipedia.org/wiki/University_of_Houston>.

The OpenARC compiler[12] <http://en.wikipedia.org/wiki/OpenACC#cite_note-12> is
an open source, C language
<http://en.wikipedia.org/wiki/C_(programming_language)> compiler developed
over the past several years at Oak Ridge National Laboratory
<http://en.wikipedia.org/wiki/Oak_Ridge_National_Laboratory> to fully
support all features in the OpenACC 1.0 specification.

An open source compiler, accULL, is also developed by the University of La
Laguna <http://en.wikipedia.org/wiki/University_of_La_Laguna> (C language
<http://en.wikipedia.org/wiki/C_(programming_language)> only).[13]
<http://en.wikipedia.org/wiki/OpenACC#cite_note-13>

GNU GCC <http://en.wikipedia.org/wiki/GNU_Compiler_Collection> supports
OpenACC starting with GCC version 5.[14]
<http://en.wikipedia.org/wiki/OpenACC#cite_note-14>[15]
<http://en.wikipedia.org/wiki/OpenACC#cite_note-15>

**Comment by PtrMan** [comment]
I am as always in a dreaming state, but what about this:

We can build a project to translate automatically a input Language (take Java) to an AST.
In a first stage the transformator can translate the AST to a target language, lets say C++ or D or Java.
Later we can add functionality for automated refactoring and or supercompilation.

**Comment by automenta** [comment]
in terms of actual results i think the most benefit will come from improvement control system designs regardless of the implementation. but to find these we may need to do intensive searches through parameter spaces driven by evolutionary algorithms and other optimization tools (the reason for the "objenome" tools). this can be easily distributed in a "nars@home" sort of crowdsourced effort. https://github.com/opennars/opennars/blob/volatile5/nars_logic/src/test/java/nars/analyze/experimental/OptimizeLeakGenetic3.java is one example for evolving a budget control function with 6 input variables. slightly overkill but illustrates an example of attempting to optimize one isolated feature that could result in a huge benefit. for example if it found that applying a X% negative bias towards derivations involving a ImageExt produced Y% cycle-sooner results as measured by test case completion, etc..

**Comment by automenta** [comment]
actually, 9 variables:

Input task priority
            derPri = leakProgram.var("derPri");
Input task quality
            derQua = leakProgram.var("derQua");
Task is a quest or question (1) or not (0)
            derQuest = leakProgram.var("derQuest");
Task is a belief (1) or not (0)
            derJudge = leakProgram.var("derJudge");
Task is a goal (1) or not (0)
            derGoal = leakProgram.var("derGoal");
Complexity of the task
            derComplex = leakProgram.var("derComplex");
Histogram of priority distribution of concept bag between bottom quarter and 1st third
            c0 = leakProgram.var("c0");
Histogram of priority distribution of concept bag between 1st third and 2nd third
            c1 = leakProgram.var("c1");
Histogram of priority distribution of concept bag between 2nd third and top
            c2 = leakProgram.var("c2");

the c0,c1,c2 give the function a sense of the current distribution of concepts in the bag which it will be attempting to insert the new task into. the output of this function is the final priority that would be assigned to the incoming task before it is input to memory.

adding additional variables for the term type and other task or system characteristics is where i think this could make some interesting distinctions.

**Comment by automenta** [comment]
example output:
https://gist.github.com/automenta/e26abe0b664145d72757

**Comment by jnorthrup** [comment]
i experimentally replaced the byte[] naming with WantzZeroCopy interfaces.  everywhere a byte[] used to be returned now returns a lambda with a bytebuffer method.  this means that all name() calls are deferred if desired, cached if desired, and will not materialize if the lambda is what produces the compound strings, until that string needs reading.

**Comment by automenta** [comment]
wow cool that might be a good way to do it. will check it out soon

**Comment by jnorthrup** [comment]
my branch is just a bridge i don't expect to see any improvements by adding
slower directmemory options until the object model is pointers to structs

**Comment by automenta** [comment]
is there any reason not to pull this WantsZeroCopy into the code now, or is it better to wait?

**Comment by jnorthrup** [comment]
give it a shake see what happens i think its a midpoint but i dont plan to
knowingly change NARS semantics just bend some interfaces over time

**Comment by jnorthrup** [comment]
with regards to bbcursive where WantsZeroCopy is defined, it is integral to
the bbcursive std.{cat,str} methods.

WantsZeroCopy provides no compareTo yet. seems cheap to wrap as a default
method.

"std.cat" needs cat(byte[]...a) operators it only knows WantZeroCopy,
byteBuffers, and ByteBufferReaders to date.  Relaxfactory was the first
bbcursive project and it never had any byte[] operators to write macros for

**Comment by automenta** [comment]
i suspect the deferred instantiation should help in cases where the term is discarded before being used in a bag, this was the idea behind the different Compound subclasses , some of which implement this halfway or never need to create a byte[] name.  but your abstraction might be cleaner than how these subclasses currently are responsible for it. anyway we can try both and see. do you see any reason not to try the int[] utf8 implementation that i started?

**Comment by jnorthrup** [comment]
why is a String not good enough here? because of heap ?  byte[] code can't
exist as struct projections on windowed mmap blocks which is where i
suspect scale is easiest to bring in

**Comment by jnorthrup** [comment]
did you have a benchmark in mind?  the drunken-wallhack consumeString()
method  parses quoted json strings and json is explicitly utf-8.

the IntBuffer could be tested in that benchmark to parse between legally
escaping json string quotes

**Comment by automenta** [comment]
a compareTo and equals() and hashCode with UTF8 int[] would theoretically be 4x faster than UTF16 string.  this is used in ordering subterms in commutative terms like Similarity.  byte[] for UTF8 saves up to 2x memory for term id's, which are used in all tasks (and sentences) and concepts.  int[] would save almost as much, costing up to 3 bytes penalty if length mod 4 bytes == 1 but i think this is acceptable for the comparison and hash benefits.

the benchmark i would use is the NALPerfLoop set to a high # of repetitions for accuracy. then comparing the results on each test and see if there is an overall benefit or if it is specific to certain tests.

**Comment by jnorthrup** [comment]
you know that all math is promoted to int right?  for base64, the byte[]
option is twice as fast as the char[] option, due strictly to io.  they
both use int promoted math to do the base64 stuff.

http://iharder.sourceforge.net/current/java/base64/

**Comment by jnorthrup** [comment]
http://migbase64.sourceforge.net/ is the impl that benches both char and
byte versions

**Comment by automenta** [comment]
maybe it wouldnt matter so much if java had an Arraycompare (wrapper for
memcmp) like it does Arraycopy. but equivalence (ie. irregardless of hash)
and compareTo ordinality involving int[x] seem like it should beat
byte[x*4] everytime. there is some code in there for an unsafe UTF8 byte[]
comparison involving an int cast but after i suspected a problem with it, i
commented it out. the pure java version of it it didnt seem to have less
performance anyway

**Comment by jnorthrup** [comment]
a subset of the jvm options
http://www.oracle.com/technetwork/articles/java/vmoptions-jsp-140102.html
-XX:+UseLargePagesUse large page memory. (Introduced in 5.0 update 5.) For
details, see Java Support for Large Memory Pages
<http://www.oracle.com/technetwork/java/javase/tech/largememory-jsp-137182.html>
.-XX:+UseMPSSUse Multiple Page Size Support w/4mb pages for the heap. Do
not use with ISM as this replaces the need for ISM. (Introduced in 1.4.0
update 1, Relevant to Solaris 9 and newer.) [1.4.1 and earlier: false]
-XX:+UseStringCacheEnables caching of commonly allocated strings.
 -XX:AllocatePrefetchLines=1Number of cache lines to load after the last
object allocation using prefetch instructions generated in JIT compiled
code. Default values are 1 if the last allocated object was an instance and
3 if it was an array.
 -XX:AllocatePrefetchStyle=1Generated code style for prefetch instructions.
0 - no prefetch instructions are generate*d*,
1 - execute prefetch instructions after each allocation,
2 - use TLAB allocation watermark pointer to gate when prefetch
instructions are executed.
 -XX:+UseCompressedStringsUse a byte[] for Strings which can be represented
as pure ASCII. (Introduced in Java 6 Update 21 Performance Release)
 -XX:+OptimizeStringConcatOptimize String concatenation operations where
possible. (Introduced in Java 6 Update 20)
 includes -XX:+UseLargePagesUse large page memory. (Introduced in 5.0
update 5.) For details, see Java Support for Large Memory Pages
<http://www.oracle.com/technetwork/java/javase/tech/largememory-jsp-137182.html>
.-XX:+UseMPSSUse Multiple Page Size Support w/4mb pages for the heap. Do
not use with ISM as this replaces the need for ISM. (Introduced in 1.4.0
update 1, Relevant to Solaris 9 and newer.) [1.4.1 and earlier: false]
-XX:+UseStringCacheEnables caching of commonly allocated strings.
 -XX:AllocatePrefetchLines=1Number of cache lines to load after the last
object allocation using prefetch instructions generated in JIT compiled
code. Default values are 1 if the last allocated object was an instance and
3 if it was an array.
 -XX:AllocatePrefetchStyle=1Generated code style for prefetch instructions.
0 - no prefetch instructions are generate*d*,
1 - execute prefetch instructions after each allocation,
2 - use TLAB allocation watermark pointer to gate when prefetch
instructions are executed.
 -XX:+UseCompressedStringsUse a byte[] for Strings which can be represented
as pure ASCII. (Introduced in Java 6 Update 21 Performance Release)
 -XX:+OptimizeStringConcatOptimize String concatenation operations where
possible. (Introduced in Java 6 Update 20)

**Comment by automenta** [comment]
me@xyz ~/s/traytention> java -XX:+UseCompressedStrings
Unrecognized VM option 'UseCompressedStrings'

this might do it but its missing in java 8

**Comment by automenta** [comment]
java -XX:+PrintFlagsFinal

**Comment by automenta** [comment]
https://github.com/opennars/opennars/blob/volatile5/doc/opennars_entities.svg this shows some of the relations between the main entities involved

i'll try making one to show the Default control feedback loop of how budget selects the premises which derive new tasks that become new input

**Comment by jnorthrup** [comment]
http://stackoverflow.com/a/15840307 answer is consistent with what i would
attack _last_, which is once bags are serialized to byte-extents then the
runtime kb state would trie-ify the keys along the lines of HAT-TRIES for
immutable strings(as utf8).  at least where it turned out that memory
access and context/fetch overheads are pathological after all other
profilling unhides it.

**Comment by jnorthrup** [comment]
WantsZeroCopy was just a practice run i don't see it bering a feature that's distinct against prautobeans implicit zero-copy options and lazy iteration.

what comes next is to convert the class hierarchy 
  
from 
![image](https://cloud.githubusercontent.com/assets/73514/7929537/131543c8-08b4-11e5-9a19-c2c7b9ac5cbb.png)

and to 

![image](https://cloud.githubusercontent.com/assets/73514/7929527/fc790f8c-08b3-11e5-91d5-ed10a96239c2.png)


into a src/main/proto/nars.proto with generator.  then the same places that were modified for WantsZeroCopy key maniuplation will need to support the generated interfaces.  semantics should not change, but i have seen at least one static singleton map that will provoke a discussion of re-entrant practices

**Comment by automenta** [comment]
some results, cycle speed seems to have improved from the new Identifier
implementation

further gains for internal identifier processing seem possible by removing
the spaces and representing all NAL operators as one character special
symbols.

/home/me/share/opennars/nal/test5/nal5.0.nal: : 86.441ms/test, -63702.527
kb/test, *2.821 uS/cycle*, 10214.0 cycles/test
/home/me/share/opennars/nal/test5/nal5.1.nal: : 157.59ms/test, 25660.281
kb/test, *5.023 uS/cycle*, 10457.0 cycles/test
/home/me/share/opennars/nal/test5/nal5.10.nal: : 149.849ms/test, -41994.766
kb/test, *4.924 uS/cycle*, 10145.0 cycles/test
/home/me/share/opennars/nal/test5/nal5.11.nal: : 159.41ms/test, 18223.73
kb/test, *5.184 uS/cycle*, 10251.0 cycles/test
/home/me/share/opennars/nal/test5/nal5.12.nal: : 95.187ms/test, 14318.875
kb/test, 3.146 uS/cycle, 10086.0 cycles/test
/home/me/share/opennars/nal/test5/nal5.13.nal: : 121.945ms/test, -47387.438
kb/test, 4.013 uS/cycle, 10129.0 cycles/test
/home/me/share/opennars/nal/test5/nal5.14.nal: : 102.851ms/test, 17528.281
kb/test, 3.402 uS/cycle, 10078.0 cycles/test
/home/me/share/opennars/nal/test5/nal5.15.nal: : 85.258ms/test, 12519.77
kb/test, 2.74 uS/cycle, 10372.0 cycles/test
/home/me/share/opennars/nal/test5/nal5.16.nal: : 127.144ms/test, 22929.312
kb/test, 4.038 uS/cycle, 10496.0 cycles/test
/home/me/share/opennars/nal/test5/nal5.17.nal: : 114.883ms/test, -37550.234
kb/test, 3.698 uS/cycle, 10356.0 cycles/test
/home/me/share/opennars/nal/test5/nal5.18.1.nal: : 95.872ms/test, 10450.812
kb/test, 3.026 uS/cycle, 10560.0 cycles/test
/home/me/share/opennars/nal/test5/nal5.18.2.nal: : 87.685ms/test, 12773.215
kb/test, 2.906 uS/cycle, 10058.0 cycles/test
/home/me/share/opennars/nal/test5/nal5.18.3.nal: : 100.188ms/test, 9290.609
kb/test, 3.219 uS/cycle, 10375.0 cycles/test
/home/me/share/opennars/nal/test5/nal5.18.4.nal: : 105.504ms/test,
17930.738 kb/test, 3.403 uS/cycle, 10335.0 cycles/test
/home/me/share/opennars/nal/test5/nal5.18.nal: : 134.105ms/test, 19051.367
kb/test, 4.238 uS/cycle, 10547.0 cycles/test
/home/me/share/opennars/nal/test5/depr/nal5.19.nal: : 107.836ms/test,
19439.734 kb/test, 3.411 uS/cycle, 10539.0 cycles/test
/home/me/share/opennars/nal/test5/nal5.2.nal: : 128.916ms/test, 24838.094
kb/test, 4.198 uS/cycle, 10236.0 cycles/test
/home/me/share/opennars/nal/test5/nal5.20.nal: : 83.279ms/test, 5196.285
kb/test, 2.694 uS/cycle, 10304.0 cycles/test
/home/me/share/opennars/nal/test5/nal5.21.nal: : 107.98ms/test, 11430.598
kb/test, 3.477 uS/cycle, 10353.0 cycles/test
/home/me/share/opennars/nal/test5/nal5.22.nal: : 71.338ms/test, 7274.375
kb/test, 2.318 uS/cycle, 10257.0 cycles/test
/home/me/share/opennars/nal/test5/nal5.23.nal: : 100.059ms/test, -34401.922
kb/test, 3.186 uS/cycle, 10467.0 cycles/test
/home/me/share/opennars/nal/test5/nal5.24.nal: : 84.691ms/test, 14046.887
kb/test, 2.722 uS/cycle, 10371.0 cycles/test
/home/me/share/opennars/nal/test5/nal5.25.1.nal: : 84.936ms/test, 15050.227
kb/test, 2.816 uS/cycle, 10053.0 cycles/test
/home/me/share/opennars/nal/test5/nal5.25.nal: : 84.835ms/test, 14514.5
kb/test, 2.784 uS/cycle, 10157.0 cycles/test
/home/me/share/opennars/nal/test5/nal5.26.nal: : 111.028ms/test, -32310.324
kb/test, 3.673 uS/cycle, 10077.0 cycles/test
/home/me/share/opennars/nal/test5/nal5.27.nal: : 109.6ms/test, 16771.738
kb/test, 3.519 uS/cycle, 10382.0 cycles/test
/home/me/share/opennars/nal/test5/nal5.28.nal: : 111.241ms/test, -31039.051
kb/test, 3.542 uS/cycle, 10469.0 cycles/test
/home/me/share/opennars/nal/test5/nal5.29.nal: : 99.578ms/test, 14417.066
kb/test, 3.156 uS/cyc

**Comment by jnorthrup** [comment]
foo.split( " *--> *" ) wfm

String[] is as good a key imho as String

**Comment by automenta** [comment]
we can try them all out as subclasses of Identifier. they just need to
implement a few methods and use consistent types between similarly typed
objects.

i've identified several differnent methods of hash, equality, and compareTo
testing for the various entities in NARS.   the two main kinds are either
involving generation of a string-like token (as is represented in byte[]
currently) or by directly comparing the data components of the entities
involved.

for example, sometimes it seems better for a 2-subterm compound to avoid
generating a byte[] name (as long as possible or indefinitely) and instead
just compare the 2 subterms against the other 2-subterm compound.

hashing is usually unavoidable once an item must be bagged, but it can be
calculated by both these 2 methods too, in 2 diferent ways.  (multiple
hashing functions should help with a customized version of CuckooHashmap,
which must hash up to 3 times for get and put.  if it can use a
pre-existing set of hashes instead of its custom hash function that lack
entropy since it only uses the original hashcode.. then it should be faster
and hash better).  even though cuckoohash put and get are slower than
hashmap, i continue to see its benefits of not allocating hash nodes.

btw do you know of a HashMap like collection which recycles its hash nodes
in a pool? i might write one using my DDList doubly-linked list collection
i use in bags. DDList is a linked list where each cell holds a reference to
some item. i use these as values in the index hashmap of a bag. this allows
both sequential cyclical prioritized access and direct removal by ID. i
pool these DD cells into an arraydeque and no allocation has to happen
(bags are generally of fixed size).

https://github.com/opennars/opennars/blob/volatile5/nars_logic/src/main/java/nars/bag/impl/experimental/ChainBag.java

heuristics to choose all the parameters of these conditions and thresholds
for these policies can be tuned by parameter search to minimize cpu,
memory, or etc..

**Comment by jnorthrup** [comment]
regarding off-heap:

1.5.5 branch "at62classes" has a full prautobeans state implementation expressing the fundamental changes required for refactoring.

first: convert all instance methods with stateful objects to static.
second: convert to structs cheaply one at a time using proto in nars_data

regarding boost fusion:

project bbcursive is using variadic template generic systeax toi define signalling tuples from immutable input bytes allowing for polylglot access to memory by default via mmap and other trampoline buffer options.
